api/samples/statecmp.c

```diff
@@ -0,0 +1,180 @@
+/* **********************************************************
+ * Copyright (c) 2021 Google, Inc.  All rights reserved.
+ * **********************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL GOOGLE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* Illustrates use of the drstatecmp extension by a sample client.
+ * This sample client introduces an instrumentation bug that is caught by
+ * drstatecmp. This sample client also shows how to specify an user-defined
+ * callback to be invoked on state comparison mismatches.
+ */
+
+#include "dr_api.h"
+#include "drreg.h"
+#include "drstatecmp.h"
+#include <string.h>
+#include <limits.h>
+
+#define MINSERT instrlist_meta_preinsert
+
+static int error_detected = 0;
+static int global_count = 0;
+
+static void
+event_exit(void)
+{
+    drreg_exit();
+    drstatecmp_exit();
+    DR_ASSERT(error_detected == 1);
+}
+
+/* Invoked by drstatecmp when a state comparison fails. */
+static void
+error_callback(const char *msg, void *tag)
+{
+    error_detected = 1;
+    DR_ASSERT_MSG(!strcmp(msg, "xflags"), msg);
+}
+
+static bool
+bb_may_have_side_effects(void *drcontext, instrlist_t *bb)
```

Comment:
Just to clarify, no other client will need to duplicate drstatecmp internal logic to use drstatecmp. This function does not specify where checks are enabled but it is just an artifact of creating a buggy client that does not lead to a flaky test. This client introduces clobbering of aflags with its instrumentation. To avoid a seg fault or any form of weird behavior, I had to restrain the client to only introduce bugs in basic blocks that drstatecmp could test.  
That said, it is worth exporting this functionality to avoid the redundancy here (and getting out-of-sync with the test client). I also renamed it to drstatecmp_bb_checks_enabled() for future-proofing as suggested.